### PR TITLE
Don't reregister subfixture finalizer in parent fixture if value is cached

### LINKED ISSUE #12135: Finalizer re-registered in parent fixture even when the value of the fixture is cached
When getting the value of a fixture, it will always add its finalizer to parent fixtures, regardless of if it's cached or not. When we have several subfixtures, this leads to unpredictable ordering between them on when they're torn down (if triggered by the parent fixture being torn down). It is also an obvious inefficiency, where a long-lived parent fixture could rack up tons of irrelevant fixtures.

As found in discussion of #11833, which resolved a similar issue. Also related to #4871

### repro
```python
import pytest


@pytest.fixture(scope="module")
def fixture_1(request):
    ...

@pytest.fixture(scope="module")
def fixture_2(fixture_1):
    print("setup 2")
    yield
    print("teardown 2")

@pytest.fixture(scope="module")
def fixture_3(fixture_1):
    print("setup 3")
    yield
    print("teardown 3")

def test_1(fixture_2):
    ...
def test_2(fixture_3):
    ...

# This will add a second copy of fixture_2's finalizer in the parent fixture, causing it to
# be torn down before fixture 3.
def test_3(fixture_2):
    ...


# Trigger finalization of fixture_1, otherwise the cleanup would sequence 3&2 before 1 as normal.
@pytest.mark.parametrize("fixture_1", [None], indirect=["fixture_1"])
def test_4(fixture_1):
    ...
```
### output
```
setup 2
setup 3
teardown 2
teardown 3
```
but if we remove `test_3` we get 2-3-3-2.

### suggested fix
Here's the relevant method:
https://github.com/pytest-dev/pytest/blob/2607fe8b4706fa701925db50f7892ddff6ed2928/src/_pytest/fixtures.py#L1037-L1079

I thought this would be a trivial fix, merely moving the code that adds finalizers to parent fixtures (L1041 to L1044) to after the check on whether the value is cached (L1062). But this broke tests in very weird ways... and I think that's the same problem as noted in https://github.com/pytest-dev/pytest/issues/4871#issuecomment-2002622725 - see "Possible solution - make the cache key not match". Current behaviour requires that the call to `request._get_active_fixturedef(argname)` happens before checking if the value is cached, to let the parent fixture check if it has been differently parametrized -> if so tear itself down -> tear us down -> invalidate our cache.

Once I figured out that the finalize-adding code had dual purposes, it was fairly easy to split it and achieve the wanted behavior without breaking anything:

```python
    def execute(self, request: SubRequest) -> FixtureValue:
        # Ensure arguments (parent fixtures) are loaded. If their cache has been
        # invalidated they will also finalize us and invalidate our cache.
        # If/when parent fixture parametrization is included in our cache key
        # this can be moved after checking our cache key.
        parent_fixtures_to_add_finalizer = []
        for argname in self.argnames:
            fixturedef = request._get_active_fixturedef(argname)
            if not isinstance(fixturedef, PseudoFixtureDef):
                # save fixture as one to add our finalizer to, if we're not cached
                # resolves #12135
                parent_fixtures_to_add_finalizer.append(fixturedef)

        my_cache_key = self.cache_key(request)
        if self.cached_result is not None:
            ...

        finalizer = functools.partial(self.finish, request=request)
        # add finalizer to parent fixtures
        for parent_fixture in parent_fixtures_to_add_finalizer:
            parent_fixture.addfinalizer(finalizer)
```

### possible bad outcomes
... I have a very hard time coming up with any downside of implementing this. Will write a PR once I've rewritten the repro as a proper regression test.

### PR DESCRIPTION
**Pull request by jakkdl**
Fixes #12135

> When getting the value of a fixture, it will always add its finalizer to parent fixtures, regardless of if it's cached or not. When we have several subfixtures, this leads to unpredictable ordering between them when they're torn down (if triggered by the parent fixture being torn down). It is also an obvious inefficiency, where a long-lived parent fixture could rack up tons of irrelevant fixtures.

Adds two regression tests: one that checks the ordering failure, and one that checks the number of finalizers.

The implementation would be slightly cleaner if parent fixture parametrization was included in the cache key, as discussed in https://github.com/pytest-dev/pytest/issues/4871#issuecomment-2002622725

### CODE REVIEW

**Comment by WarrenTheRabbit** [comment]
@jakkdl: I like how you add context by including the related issue in the source code of your regression test. I think I'll adopt that practice myself! Is it a common policy?

Other than a point of praise, I don't really have anything to say. _pytest_ is still quite opaque to me. But I'll read through the issues you cite (thank you for including them) when I have time and see if I can follow along with your approach.

**Comment by jakkdl** [comment]
> @jakkdl: I like how you add context by including the related issue in the source code of your regression test. I think I'll adopt that practice myself! Is it a common policy?

I have seen different versions of it in different repos, so I'd say it's relatively common. This repo also has a couple files dedicated to being regression tests for issues, e.g.
```
testing/example_scripts/issue88_initial_file_multinodes
testing/example_scripts/issue_519.py
testing/examples/test_issue519.py
```
There's also stuff like https://docs.astral.sh/ruff/rules/missing-todo-link/
 
> Other than a point of praise, I don't really have anything to say. _pytest_ is still quite opaque to me. But I'll read through the issues you cite (thank you for including them) when I have time and see if I can follow along with your approach.

Oh yeah it took me a while to grok the fixture infrastructure, not for the faint of heart!

**Comment by WarrenTheRabbit** [comment]
> This repo also has a couple files dedicated to being regression tests for issues

> There's also stuff like https://docs.astral.sh/ruff/rules/missing-todo-link/

Thanks for the thoughtful response. _ruff_ looks cool too.

> Oh yeah it took me a while to grok the fixture infrastructure, not for the faint of heart!

Ha! I'm at the point where if I say anything at all about the fixture infrastructure I invariably say nonsensical things and embarrass myself horribly. But am definitely trying to be a citizen of this place - one fainting spell at a time. In fact, wrote some things about it the other day in an Issue and likely face-planted myself all over the pavement. Such is the journey I suppose. Onwards and upwards. But also: _Eeep!_

I won't yatter on and make it harder for someone to get up to speed on your PR. Thanks, @jakkdl!

**Review by bluetech — CHANGES_REQUESTED** [review]
Thanks @jakkdl! As mentioned in the issue, this change LGTM. I left some comments mostly on terminology.

File: src/_pytest/fixtures.py [hunk 1]
```diff
@@ -1035,13 +1035,18 @@ def finish(self, request: SubRequest) -> None:
...
 
     def execute(self, request: SubRequest) -> FixtureValue:
-        finalizer = functools.partial(self.finish, request=request)
-        # Get required arguments and register our own finish()
-        # with their finalization.
+        # Ensure arguments (parent fixtures) are loaded.
+        # If their cache has been invalidated it will finish itself and subfixtures,
+        # which will set our self.cached_result = None.
+        # If/when parent fixture parametrization is included in our cache key this
+        # can be moved after checking our cache key and not require saving in a list.
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply src/_pytest/fixtures.py:1042]: > XXX: figure out why this is needed even in the cached case

The reason this is needed in the cached case is also because of the requested-fixture-parametrization case (and possible other cases where the requested fixture's cache is invalidated?). This fixture's cache key does not change depending on the cache key of the requested fixture, and the following couple lines only check our cache key before deciding whether to return early.

> Also make it such that whenever a fixture we request is torn down, this fixture is torn down first.

I slightly disagree on putting this up top instead of when this is actually ensured, where you wanted to remove the comment, but I'll reword it slightly different to not directly imply that this is a direct consequence of the lines directly following this.

File: changelog/12135.bugfix.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix subfixtures adding their finalizer multiple times to parent fixtures, causing incorrect teardown ordering in some instances.
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply changelog/12135.bugfix.rst]: "a fixture that this fixture requests" turned out to be very wordy and hard to read, so I swapped to using "dependent fixtures" - which seems to be somewhat established and shouldn't be possible to interpret as the reverse.

Switched to non-intuitive + unreliable (since teardown order will change depending on which fixture was last requested)

File: src/_pytest/fixtures.py [hunk 3]
```diff
@@ -1060,6 +1065,11 @@ def execute(self, request: SubRequest) -> FixtureValue:
             self.finish(request)
             assert self.cached_result is None
 
+        finalizer = functools.partial(self.finish, request=request)
+        # add finalizer to parent fixtures
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply src/_pytest/fixtures.py]: I removed it.. and added it back. Feel free to remove it again if you still think it is detracting

File: src/_pytest/fixtures.py [hunk 4]
```diff
@@ -1035,14 +1035,25 @@ def finish(self, request: SubRequest) -> None:
             raise BaseExceptionGroup(msg, exceptions[::-1])
 
     def execute(self, request: SubRequest) -> FixtureValue:
-        finalizer = functools.partial(self.finish, request=request)
-        # Get required arguments and register our own finish()
-        # with their finalization.
+        """Return the value of this fixture, executing it if not cached."""
+        # Ensure dependent fixtures that this fixture requests are loaded.
```
Comment (bluetech) [inline src/_pytest/fixtures.py:1039]: This comment has a grammar issue, can you rephrase it?

Comment (jakkdl) [reply src/_pytest/fixtures.py]: I failed to see the problem, but Claude fixed it :)

**Review by bluetech — APPROVED** [review]

**Comment by bluetech** [comment]
Thanks @jakkdl!

BTW, I'm going to send you an invite to join pytest-dev. This doesn't require anything from you really, but will allow you to edit issues, merge your PRs, etc. if you'd like to contribute more.

**Comment by jakkdl** [comment]
> Thanks @jakkdl!
> 
> BTW, I'm going to send you an invite to join pytest-dev. This doesn't require anything from you really, but will allow you to edit issues, merge your PRs, etc. if you'd like to contribute more.

Sweet, might come in handy!
